tripal/t4d8
# Custom chado install (supports multi-chado)

Issue #57 
- Upgrades the Chado installer form
- Provides a Chado installer as a Drupal Service
    - Supports installing in different schema names
    - Uses a new install method applying the SQL to the database directly in chunks
    - Includes the script to create the chunk files
- Provides a Generic SQL/Schema File import Drupal Service
   - I abstracted most of the functionality from the chado install to be used for alternate postgreSQL databases
   - useful for alternate data backends as well as just populating your chado database.
 - Upgrades the ChadoSchema API Class
   - includes all Tripal3 functionality plus a few more methods
   - includes backwards compatible procedural API
 - Includes Documentation (see screenshots)
   - developer docs for both services
   - generic docs supporting alternate data storage backends

## Notes
 - Job system is not upgraded so chado install form just provides a drush command.
 - Install is backwards compatible and both `tripal_chado_load_drush_submit` and `tripal_chado_install_chado` are still supported.
 - Backwards compatible schema api using procedural functions is also provided and tested.

## Automated Testing
   - requires a `testchado` schema pre-installed with chado
   - I recommend not having a `chado` named schema while testing
   - Chado Install Form: chadoInstallerFormTest
   - Chado Install: chadoInstallerTest
        - Can't test easily due to different use for database connections.
        - currently tests testchado install since it had to be done with the service
   - Schema API: SchemaAPITest
   - Chado compliance test: ChadoComplianceTest
      - Tests that the chado installed matches the schema api
      - Great tests for ensuring any bugs are not related to deviating from the schema  

##  Manual Testing
- Form is at admin/tripal/storage/chado/install
- Requires drush
- schema name should not matter and you can install multiple copies of chado

## Screenshots
![Firefox_Screenshot_2020-07-11T18-02-36 731Z](https://user-images.githubusercontent.com/1566301/87230777-3ff49080-c36f-11ea-89b3-815880074dfd.png)
